### ISSUE TITLE
API stability, plus docs and examples?

### ISSUE THREAD

**Issue by bdewilde** [explosion/thinc#21]
I've been hemming and hawing about adding pre-trained/trainable NN models into `textacy`, partly because I've been waiting to see how `thinc` develops rather than committing early to, say, `keras`. Do you have any sense of when `thinc`'s NN API will be stable? And do you plan to add more documentation and examples? [Your latest blog post](https://explosion.ai/blog/supervised-similarity-siamese-cnn) is great, but the code seems to very much be a work-in-progress.

As always, thanks for all your great work. I look forward to seeing how this package matures!

**Comment by honnibal** [comment]
If you're trying to get stuff done I think you'll be best off using PyTorch for now. I think it's already annoying enough to wonder whether an NN model isn't training because of the data or the architecture, without worrying that the underlying implementation simply doesn't do what's promised :p 

The core NN code is short in all of these frameworks, so if you like Thinc later you can always port your model.

**Comment by adam-ra** [comment]
This library looks very exciting, something close to no-sugar-added NN package ready for tagging problems! I understand that you're overwhelmed with tasks around Spacy but if you found time to add in example training file (tiny snippet) for a tagger, I would appreciate it greatly!
Unless you think that for making custom taggers and entity recognisers it's better to use Spacy and train its models without the need to stepping down into this black box :)

**Comment by honnibal** [comment]
Have you seen the `examples/cnn_tagger.py` file? It uses the AnCora corpus, but you could apply the same code to any UD treebank.

I think it does make sense to go straight to the neural network approach for these problems. The deep CNN is particularly interesting for tagging imo.

A researcher at UC Maryland who wrote a nice paper on dilated CNN for NER points out that this is basically a joint model over the whole sentence. I hadn't thought of that perspective, but it's an interesting insight. It means that the step-through model in spaCy's tagger class my be unnecessary. We don't need to have features for the previously decided tags, since the representation used to compute those tags is available as an even better representation.

**Comment by adam-ra** [comment]
Oh, I've missed the thinc.extra.datasets in this example, thanks!

I'll look up the paper, it sounds indeed interesting! It's hard for me to imagine any representation that could handle sentences of any size without sliding window, which makes me even more eager to read on that. I have some very good experience with using an array of CRFs to handle each tagset attribute as CRFs also do well with longer-distance dependencies.

**Comment by honnibal** [comment]
You could call it a sliding window. Let's say you're taking the input and labels:

```
a a a a a a a a a
0 0 1 0 0 1 0 0 1
```

Every third symbol in the input receives a "1", otherwise the output is "0". Obviously so long as we have two states of history, a markov model can learn this easily. It's easy to imagine how the LSTM would work for this too.

The CNN isn't as intuitive. Consider the first layer:

    inputs = concat(prev_input[i-1], prev_input[i], prev_input[i+1]) 

If the previous input is the original string, the representation for this first layer is the same almost everywhere --- it's only different at the beginning and end. So we can distinguish three "states" here: beginning, middle and end:
    
```
a a a a a a a a a
b m m m m m m e
0 0 1 0 0 1 0 0 1
```

But now when we do the next layer, we're making some progress. After the next convolution, we could learn new states for second and paenultimate (second last):

```
a a a a a a a a a
b m m m m m m e
b s m m m m p e
0 0 1 0 0 1 0 0 1
```

Now we finally have some unambiguous features. Our input-to-output map for the next layer looks like this:

```
(_, b, s): 0
(b, s, m): 0
(s, m, m): 1
(m, m, m): 0
(m, m, p): 1
(m, p, e): 0
(p, e, _): 0
```

Each layer of depth allows information to flow from two more tokens (one on either side) --- so it increases the window size by 1.

**Comment by honnibal** [comment]
Three years later...Stay tuned ;). We've had a massive sprint on this over the last month or so, and we're close to a release.

**Comment by bdewilde** [comment]
@honnibal I've been watching progress on the `develop` branch... Very excited! Thanks for all your hard work.
